payjoin/src/core/receive/mod.rs

```diff
@@ -749,25 +772,117 @@ pub(crate) mod tests {
         };
-        let p2tr_pair = InputPair::new_p2tr(p2tr_txout.clone(), outpoint, Some(sequence)).unwrap();
-        assert_eq!(p2tr_pair.txin.previous_output, outpoint);
-        assert_eq!(p2tr_pair.txin.sequence, sequence);
-        assert_eq!(p2tr_pair.psbtin.witness_utxo.unwrap(), p2tr_txout);
+
+        // Expected weight for p2tr keyspend
+        let expected_key_weight =
+            InputWeightPrediction::P2TR_KEY_DEFAULT_SIGHASH.weight() + NON_WITNESS_DATA_WEIGHT;
+        let keyspend_pair =
+            InputPair::new_p2tr_keyspend(p2tr_txout.clone(), outpoint, Some(sequence)).unwrap();
+        assert_eq!(keyspend_pair.txin.previous_output, outpoint);
+        assert_eq!(keyspend_pair.txin.sequence, sequence);
+        let witness_utxo = keyspend_pair.psbtin.witness_utxo.clone().unwrap();
+        assert_eq!(witness_utxo, p2tr_txout);
+        assert_eq!(keyspend_pair.expected_weight, expected_key_weight);
+
+        // Expected weight for p2tr scriptpath
+        let script_expected_weight = Weight::from_wu(2048);
+        let script_pair = InputPair::new_p2tr_scriptpath_spend(
+            p2tr_txout.clone(),
+            outpoint,
+            Some(sequence),
+            script_expected_weight,
+        )
+        .unwrap();
+        assert_eq!(script_pair.expected_weight, script_expected_weight);
+
+        // Unsupported weight with no witness
+        let txin = TxIn { previous_output: outpoint, sequence, ..Default::default() };
+        let psbtin = psbt::Input { witness_utxo: Some(witness_utxo.clone()), ..Default::default() };
+        let p2tr_pair = InputPair::new(txin.clone(), psbtin.clone(), None);
         assert_eq!(
-            p2tr_pair.expected_weight,
-            InputWeightPrediction::P2TR_KEY_DEFAULT_SIGHASH.weight() + NON_WITNESS_DATA_WEIGHT
+            p2tr_pair.err().unwrap(),
+            PsbtInputError::from(InternalPsbtInputError::from(InputWeightError::NotSupported))
         );
 
+        // Manual expected weight with no witness
+        let manual_weight = Weight::from_wu(2048);
+        let p2tr_pair = InputPair::new(txin, psbtin, Some(manual_weight)).unwrap();
+        assert_eq!(p2tr_pair.expected_weight, manual_weight);
+
+        // If UTXO is non-P2TR
         let p2sh_txout = TxOut {
             value: Default::default(),
             script_pubkey: ScriptBuf::new_p2sh(&ScriptHash::all_zeros()),
         };
         assert_eq!(
-            InputPair::new_p2tr(p2sh_txout, outpoint, Some(sequence)).err().unwrap(),
+            InputPair::new_p2tr_keyspend(p2sh_txout, outpoint, Some(sequence)).err().unwrap(),
             PsbtInputError::from(InvalidScriptPubKey(AddressType::P2tr))
         )
     }
 
+    #[test]
+    fn p2tr_expected_weight_from_witness() {
+        let outpoint = OutPoint { txid: Txid::from_byte_array(DUMMY32), vout: 31 };
+        let sequence = Sequence::from_512_second_intervals(123);
```

Comment:
```suggestion
        let sequence = Sequence::ZERO;
```
Nit: a bit cleaner
https://docs.rs/bitcoin/0.32.8/bitcoin/struct.Sequence.html#associatedconstant.ZERO